### ISSUE TITLE
[15→16] sale_purchase: AttributeError: 'ir.model.constraint' object has no attribute '_module_data_uninstall'

### ISSUE THREAD

**Issue by pjrobertson** [OCA/OpenUpgrade#4420]
Upgrading from Odoo 15.0 to 16.0 using Open upgrade, I get the following error:

```
2024-05-01 05:46:52,621 39069 CRITICAL odoo_15 odoo.service.server: Failed to initialize database `odoo_15`. 
Traceback (most recent call last):
  File "./src/odoo/service/server.py", line 1299, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "<decorator-gen-16>", line 2, in new
  File "./src/odoo/tools/func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "./src/odoo/modules/registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "./src/odoo/modules/loading.py", line 484, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "./src/odoo/modules/loading.py", line 372, in load_marked_modules
    loaded, processed = load_module_graph(
  File "./src/odoo/modules/loading.py", line 236, in load_module_graph
    migrations.migrate_module(package, 'post')
  File "./openupgrade_16/openupgrade_framework/odoo_patch/odoo/modules/migration.py", line 18, in migrate_module
    MigrationManager.migrate_module._original_method(self, pkg, stage)
  File "./src/odoo/modules/migration.py", line 189, in migrate_module
    migrate(self.cr, installed_version)
  File "XXX/site-packages/openupgradelib/openupgrade.py", line 2281, in wrapped_function
    func(
  File "./openupgrade_16/openupgrade_scripts/scripts/sale_purchase/16.0.1.0/post-migration.py", line 83, in migrate
    constraint._module_data_uninstall()
AttributeError: 'ir.model.constraint' object has no attribute '_module_data_uninstall'
```

## Module
sale_purchase

## Describe the bug
Using latest `openupgrade_16`, and the following command:
`src/odoo-bin -c odoo16.conf --database=odoo_15 --upgrade-path=./openupgrade_16/openupgrade_scripts/scripts --load=base,web,openupgrade_framework --update all --stop-after-init`

I put a breakpoint in `sale_purchase/16.0.1.0/post-migration.py(84)` and tried to debug, not sure if it's helpful:
A `git blame` on that file leads to [this commit,](https://github.com/OCA/OpenUpgrade/commit/5441fe7115f877063a7bb46a3de53d491f7a5f9e) which is the first commit to enable support for sale_purchase

```
(Pdb) constraint
ir.model.constraint(915,)
(Pdb) constraint.id
(Pdb) constraint.type
'u'
(Pdb) constraint.name
'product_template_service_to_purchase'
(Pdb) dir(constraint)
['CONCURRENCY_CHECK_FIELD', '_BaseModel__base_classes', '_BaseModel__ensure_xml_id', '__add__', '__and__', '__bool__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__int__', '__iter__', '__last_update', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__nonzero__', '__or__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__setattr__', '__setitem__', '__sizeof__', '__slots__', '__str__', '__sub__', '__subclasshook__', '_abstract', '_active_name', '_add_fake_fields', '_add_field', '_add_inherited_fields', '_add_missing_default_values', '_add_precomputed_values', '_add_sql_constraints', '_alias_get_error_message', '_allow_sudo_commands', '_apply_ir_rules', '_auto', '_auto_init', '_build_model', '_build_model_attributes', '_build_model_check_base', '_build_model_check_parent', '_cache', '_check_company', '_check_company_auto', '_check_m2m_recursion', '_check_parent_path', '_check_qorder', '_check_recursion', '_check_removed_columns', '_compute_concurrency_field', '_compute_display_name', '_compute_field_value', '_constraint_methods', '_context', '_convert_records', '_convert_to_record', '_convert_to_write', '_cr', '_create', '_custom', '_date_name', '_depends', '_description', '_export_rows', '_extract_records', '_fetch_field', '_field_definitions', '_fields', '_fields_view_get', '_filter_access_rules', '_filter_access_rules_python', '_flush', '_flush_search', '_fold_name', '_generate_m2o_order_by', '_generate_order_by', '_generate_order_by_inner', '_get_access_action', '_get_base_lang', '_get_default_activity_view', '_get_default_calendar_view', '_get_default_form_view', '_get_default_graph_view', '_get_default_kanban_view', '_get_default_pivot_view', '_get_default_search_view', '_get_default_tree_view', '_get_external_ids', '_get_html_link', '_get_mail_redirect_suggested_company', '_get_placeholder_filename', '_get_view', '_get_view_cache', '_get_view_cache_key', '_get_view_field_attributes', '_get_view_fields', '_has_onchange', '_ids', '_in_cache_without', '_inherit', '_inherit_children', '_inherit_module', '_inherits', '_inherits_check', '_inherits_children', '_inherits_join_add', '_inherits_join_calc', '_init_column', '_init_constraints_onchanges', '_invalidate_cache', '_is_an_ordinary_table', '_load_records', '_load_records_create', '_load_records_write', '_local_sql_constraints', '_log_access', '_mail_get_message_subtypes', '_mail_track', '_mapped_func', '_message_get_default_recipients', '_model_classes', '_modified', '_modified_triggers', '_module', '_name', '_name_search', '_notify_by_email_get_headers', '_notify_get_reply_to', '_notify_get_reply_to_formatted_email', '_onchange_eval', '_onchange_methods', '_onchange_spec', '_ondelete_methods', '_order', '_origin', '_original_module', '_parent_name', '_parent_store', '_parent_store_compute', '_parent_store_create', '_parent_store_update', '_parent_store_update_prepare', '_patch_method', '_pop_field', '_populate', '_populate_dependencies', '_populate_factories', '_populate_sizes', '_prefetch_ids', '_prepare_create_values', '_prepare_setup', '_read', '_read_format', '_read_group', '_read_group_expand_full', '_read_group_fill_results', '_read_group_fill_temporal', '_read_group_format_result', '_read_group_prepare', '_read_group_prepare_data', '_read_group_process_groupby', '_read_group_raw', '_read_group_resolve_many2x_fields', '_read_progress_bar', '_rec_name', '_rec_name_fallback', '_rec_names_search', '_recompute_field', '_recompute_model', '_recompute_recordset', '_reflect_constraint', '_reflect_constraints', '_reflect_model', '_register', '_register_hook', '_revert_method', '_search', '_search_panel_domain_image', '_search_panel_field_image', '_search_panel_global_counters', '_search_panel_sanitized_parent_hierarchy', '_search_panel_selection_range', '_setup_base', '_setup_complete', '_setup_done', '_setup_fields', '_sql_constraints', '_table', '_table_has_rows', '_table_query', '_transient', '_transient_max_count', '_transient_max_hours', '_translate', '_uid', '_unregister_hook', '_update_cache', '_update_field_translations', '_valid_field_parameter', '_validate_fields', '_web_read_group', '_where_calc', '_write', 'action_archive', 'action_unarchive', 'browse', 'check_access_rights', 'check_access_rule', 'check_field_access_rights', 'clear_caches', 'concat', 'copy', 'copy_data', 'copy_translations', 'create', 'create_date', 'create_uid', 'default_get', 'definition', 'display_name', 'ensure_one', 'env', 'exists', 'export_data', 'fields_get', 'fields_get_keys', 'fields_view_get', 'filtered', 'filtered_domain', 'flush', 'flush_model', 'flush_recordset', 'get_base_url', 'get_empty_list_help', 'get_external_id', 'get_field_translations', 'get_formview_action', 'get_formview_id', 'get_import_templates', 'get_metadata', 'get_view', 'get_views', 'get_website_meta', 'get_xml_id', 'id', 'ids', 'init', 'invalidate_cache', 'invalidate_model', 'invalidate_recordset', 'is_transient', 'load', 'load_views', 'mapped', 'message', 'model', 'modified', 'module', 'name', 'name_create', 'name_get', 'name_search', 'new', 'onchange', 'pool', 'qweb_render_view', 'read', 'read_group', 'read_progress_bar', 'recompute', 'refresh', 'search', 'search_count', 'search_panel_select_multi_range', 'search_panel_select_range', 'search_read', 'sorted', 'sudo', 'toggle_active', 'type', 'union', 'unlink', 'update', 'update_field_translations', 'update_field_translations_sha', 'user_has_groups', 'view_header_get', 'web_read_group', 'web_search_read', 'with_company', 'with_context', 'with_env', 'with_prefetch', 'with_user', 'write', 'write_date', 'write_uid']
```

## To Reproduce

1. Install an Odoo 15 version, make sure sale_purchase is enabled
2. Try to upgrade to Odoo 16

**Additional context**
I am currently in the process of upgrading from Odoo 14.0. The 14.0 > 15.0 process went OK without any problems, but now 15.0 > 16.0 is throwing this error.

Earlier log messages:

```
2024-05-01 06:37:01,657 40091 INFO odoo_15 odoo.modules.loading: Loading module sale_purchase (100/112) 
2024-05-01 06:37:02,057 40091 INFO odoo_15 odoo.modules.registry: module sale_purchase: creating or updating database tables 
2024-05-01 06:37:02,185 40091 INFO odoo_15 odoo.schema: Keep unexpected index product_template_message_main_attachment_id_index on table product_template 
2024-05-01 06:37:02,185 40091 INFO odoo_15 odoo.schema: Keep unexpected index sale_order_message_main_attachment_id_index on table sale_order 
2024-05-01 06:37:02,185 40091 INFO odoo_15 odoo.schema: Keep unexpected index sale_order_date_order_index on table sale_order 
2024-05-01 06:37:02,185 40091 INFO odoo_15 odoo.schema: Keep unexpected index purchase_order_message_main_attachment_id_index on table purchase_order 
2024-05-01 06:37:02,185 40091 INFO odoo_15 odoo.schema: Keep unexpected index product_product_message_main_attachment_id_index on table product_product 
2024-05-01 06:37:02,213 40091 INFO odoo_15 odoo.modules.loading: loading sale_purchase/data/mail_templates.xml 
2024-05-01 06:37:02,236 40091 INFO odoo_15 odoo.modules.loading: loading sale_purchase/views/product_views.xml 
2024-05-01 06:37:02,272 40091 WARNING odoo_15 odoo.addons.openupgrade_framework.odoo_patch.odoo.addons.base.models.ir_ui_view: Can't render custom view sale_purchase.product_template_form_view_inherit for model product.template. Assuming you are migrating between major versions of Odoo. Please review the view contents manually after the migration.
Error: Field "property_account_creditor_price_difference" does not exist in model "product.template" 
2024-05-01 06:37:02,272 40091 WARNING odoo_15 odoo.addons.openupgrade_framework.odoo_patch.odoo.addons.base.models.ir_ui_view: Can't render custom view sale_purchase.product_template_form_view_inherit for model product.template. Assuming you are migrating between major versions of Odoo. Please review the view contents manually after the migration.
Error: Field `property_account_creditor_price_difference` does not exist 
2024-05-01 06:37:02,273 40091 INFO odoo_15 odoo.modules.loading: loading sale_purchase/views/sale_order_views.xml 
2024-05-01 06:37:02,318 40091 INFO odoo_15 odoo.modules.loading: loading sale_purchase/views/purchase_order_views.xml 
2024-05-01 06:37:02,354 40091 INFO odoo_15 odoo.modules.loading: loading sale_purchase/wizards/sale_order_cancel_views.xml 
2024-05-01 06:37:02,366 40091 INFO odoo_15 odoo.modules.migration: module sale_purchase: Running migration [16.0.1.0>] post-migration 
2024-05-01 06:37:02,379 40091 INFO odoo_15 OpenUpgrade: sale_purchase: post-migration script called with version 15.0.1.0 
2024-05-01 06:37:02,381 40091 DEBUG odoo_15 OpenUpgrade: 0 rows affected after 0:00:00.000772 running 
        insert into ir_property (
            company_id, fields_id, value_integer, name, res_id, type
        )
        select
            company_id,
            (6040),
            1,
            'service_to_purchase',
            'product.template,' || id,
            'boolean'
        from product_template
        where
            company_id is not null and
            service_to_purchase
        order by id
         
2024-05-01 06:37:02,382 40091 DEBUG odoo_15 OpenUpgrade: 0 rows affected after 0:00:00.001048 running 
        insert into ir_property (
            company_id,
            fields_id,
            value_integer,
            name,
            res_id,
            type
        )
        select
            rc.id,
            (6040),
            1,
            'service_to_purchase',
            'product.template,' || pt.id,
            'boolean'
        from product_template as pt
        inner join res_company as rc on
            pt.company_id is null and
            pt.service_to_purchase
        order by pt.id, rc.id
         
2024-05-01 06:37:02,382 40091 ERROR odoo_15 OpenUpgrade: sale_purchase: error in migration script /Users/patrick/Developer/odoo-upgrade/odoo15/openupgrade_16/openupgrade_scripts/scripts/sale_purchase/16.0.1.0/post-migration.py: 'ir.model.constraint' object has no attribute '_module_data_uninstall' 
2024-05-01 06:37:02,382 40091 ERROR odoo_15 OpenUpgrade: 'ir.model.constraint' object has no attribute '_module_data_uninstall' 
```

**Comment by pedrobaeza** [comment]
Fixed in #4422

**Comment by pjrobertson** [comment]
Brilliant! I tried commenting out the `_module_data_uninstall` line and it worked, but wasn't sure if it was the 'right' thing to do - seems like it was.

Thanks again!
